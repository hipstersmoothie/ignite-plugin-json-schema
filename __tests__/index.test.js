import _ from 'lodash';
import React from 'react';
import renderer from 'react-test-renderer';

import Schema, { ListItem, PropertyList } from '../src';

const render = node => renderer.create(node).toJSON();

describe('ListItem', () => {
  test('no value', () => {
    expect(
      render(<ListItem name="testItem" type="something" />)
    ).toMatchSnapshot();
  });

  test('with value', () => {
    expect(
      render(
        <ListItem
          name="testItem"
          type="something"
          values={['one', { title: 'two' }, 'three']}
        />
      )
    ).toMatchSnapshot();
  });

  test('with sub props', () => {
    expect(
      render(
        <ListItem
          name="testItem"
          type="object"
          properties={{
            prop1: {
              type: 'string'
            },
            prop2: {
              type: 'string'
            }
          }}
        />
      )
    ).toMatchSnapshot();
  });
});

describe('PropertyList', () => {
  test('no properties', () => {
    expect(render(<PropertyList title="Title" />)).toMatchSnapshot();
  });

  test('has properties', () => {
    expect(
      render(
        <PropertyList
          title="Title"
          items={[
            { name: 'testItem', type: 'something' },
            {
              name: 'testItem2',
              type: 'something',
              values: ['one', { title: 'two' }, 'three']
            }
          ]}
        />
      )
    ).toMatchSnapshot();
  });
});

describe('Schema Component', () => {
  const _initData = {
    title: 'title',
    id: 'root',
    description: 'description',
    required: ['requiredProp'],
    properties: {
      requiredProp: {
        type: 'string'
      },
      optionalProp: {
        type: 'string'
      }
    }
  };

  test('root def not found', () => {
    expect(render(<Schema _initData={_initData} />)).toMatchSnapshot();
  });

  test('root def found', () => {
    expect(
      render(<Schema id="root" _initData={_initData} />)
    ).toMatchSnapshot();
    expect(
      render(<Schema options={[{ args: ['root'] }]} _initData={_initData} />)
    ).toMatchSnapshot();
  });

  test('displays correct title', () => {
    expect(
      render(<Schema id="root" _initData={_.omit(_initData, 'title')} />)
    ).toMatchSnapshot();
    expect(
      render(
        <Schema
          options={[{ args: ['root'] }]}
          _initData={{
            ..._.omit(_initData, ['title', 'id']),
            $id: 'root'
          }}
        />
      )
    ).toMatchSnapshot();
  });

  test('excludes plugins', () => {
    expect(
      render(
        <Schema
          id="root"
          _initData={_initData}
          plugins={{ optionalProp: () => {} }}
          _injectedComponents={{
            optionalProp: () => <div>Custom</div>,
            nonExistentProp: () => <div>Nothing here</div>
          }}
        />
      )
    ).toMatchSnapshot();
  });

  test('excludes omitted property', () => {
    expect(
      render(
        <Schema
          id="root"
          _initData={_.omit(_initData, 'required')}
          omitProperties={['requiredProp']}
        />
      )
    ).toMatchSnapshot();
  });
});
