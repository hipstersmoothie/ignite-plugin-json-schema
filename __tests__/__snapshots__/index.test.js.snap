// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem no value 1`] = `
<li>
  <b>
    testItem
  </b>
  <i>
    [
    something
    ]
  </i>
  
</li>
`;

exports[`ListItem with sub props 1`] = `
<li>
  <b>
    testItem
  </b>
  <i>
    [
    object
    ]
  </i>
  
  <div>
    
    <ul>
      <li>
        <b>
          prop1
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
      <li>
        <b>
          prop2
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</li>
`;

exports[`ListItem with value 1`] = `
<li>
  <b>
    testItem
  </b>
  <i>
    [
    something
    ]
  </i>
  <code>
    one
    two
    three
  </code>
</li>
`;

exports[`PropertyList has properties 1`] = `
<div>
  <h4>
    Title
  </h4>
  <ul>
    <li>
      <b>
        testItem
      </b>
      <i>
        [
        something
        ]
      </i>
      
    </li>
    <li>
      <b>
        testItem2
      </b>
      <i>
        [
        something
        ]
      </i>
      <code>
        one
        two
        three
      </code>
    </li>
  </ul>
</div>
`;

exports[`PropertyList no properties 1`] = `null`;

exports[`Schema Component displays correct title 1`] = `
<div>
  <h1>
    root
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Required
    </h4>
    <ul>
      <li>
        <b>
          requiredProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
  <div>
    <h4>
      Optional
    </h4>
    <ul>
      <li>
        <b>
          optionalProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Schema Component displays correct title 2`] = `
<div>
  <h1>
    root
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Required
    </h4>
    <ul>
      <li>
        <b>
          requiredProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
  <div>
    <h4>
      Optional
    </h4>
    <ul>
      <li>
        <b>
          optionalProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Schema Component excludes omitted property 1`] = `
<div>
  <h1>
    title
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Optional
    </h4>
    <ul>
      <li>
        <b>
          optionalProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Schema Component excludes plugins 1`] = `
<div>
  <h1>
    title
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Required
    </h4>
    <ul>
      <li>
        <b>
          requiredProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
  <div>
    Custom
  </div>
  <div>
    Nothing here
  </div>
</div>
`;

exports[`Schema Component root def found 1`] = `
<div>
  <h1>
    title
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Required
    </h4>
    <ul>
      <li>
        <b>
          requiredProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
  <div>
    <h4>
      Optional
    </h4>
    <ul>
      <li>
        <b>
          optionalProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Schema Component root def found 2`] = `
<div>
  <h1>
    title
  </h1>
  <p>
    description
  </p>
  <h2>
    Structure
  </h2>
  <div>
    <h4>
      Required
    </h4>
    <ul>
      <li>
        <b>
          requiredProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
  <div>
    <h4>
      Optional
    </h4>
    <ul>
      <li>
        <b>
          optionalProp
        </b>
        <i>
          [
          string
          ]
        </i>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Schema Component root def not found 1`] = `<div />`;
